modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/GlobalSnapshotStateChannelEventsProcessor.scala

```diff
@@ -67,26 +66,63 @@ object GlobalSnapshotStateChannelEventsProcessor {
+              case Some(snapshot) =>
+                Async[F].delay {
+                  val stakingBalance = fetchStakingBalance(event.address, lastGlobalSnapshotInfo)
+                  val maybeOwnerAddress = snapshot.value.messages.flatMap(_.find(_.messageType == MessageType.Owner)).map(_.address)
+                  val maybeStakingAddress = snapshot.value.messages.flatMap(_.find(_.messageType == MessageType.Staking)).map(_.address)
+                  SnapshotFeesInfo(allFeesAddresses, stakingBalance, maybeOwnerAddress, maybeStakingAddress)
+                }
+            }
+        }
+
       def process(
         snapshotOrdinal: SnapshotOrdinal,
         lastGlobalSnapshotInfo: GlobalSnapshotInfo,
         events: List[StateChannelOutput],
         validationType: StateChannelValidationType
-      )(implicit hasher: Hasher[F]): F[StateChannelAcceptanceResult] =
-        events.traverse { event =>
-          val stakingBalance = fetchStakingBalance(event.address, lastGlobalSnapshotInfo)
+      )(implicit hasher: Hasher[F]): F[StateChannelAcceptanceResult] = {
+        val allFeesAddresses: Map[Address, Set[Address]] = getFeeAddresses(lastGlobalSnapshotInfo)
+        type Acc = (Map[Address, Set[Address]], List[ValidatedNec[(Address, StateChannelValidationError), StateChannelOutput]])
 
-          (validationType match {
-            case StateChannelValidationType.Full =>
-              stateChannelValidator.validate(event, snapshotOrdinal, stakingBalance)
-            case StateChannelValidationType.Historical =>
-              stateChannelValidator.validateHistorical(event, snapshotOrdinal, stakingBalance)
-          }).map(_.errorMap(error => (event.address, error)))
-        }
-          .map(_.partitionMap(_.toEither))
-          .flatTap {
-            case (invalid, _) => logger.warn(s"Invalid state channels events: $invalid").whenA(invalid.nonEmpty)
+        events
+          .foldLeftM[F, Acc]((allFeesAddresses, List.empty)) {
+            case ((prevAllFeeAddresses, alreadyProcessed), event) =>
+              buildSnapshotFeesInfo(lastGlobalSnapshotInfo, event, prevAllFeeAddresses).flatMap { snapshotFeesInfo =>
+                val validationV = validationType match {
+                  case StateChannelValidationType.Full =>
+                    stateChannelValidator.validate(event, snapshotOrdinal, snapshotFeesInfo)
+                  case StateChannelValidationType.Historical =>
+                    stateChannelValidator.validateHistorical(event, snapshotOrdinal, snapshotFeesInfo)
+                }
+
+                validationV.map {
+                  case valid @ Validated.Valid(event) =>
+                    val updatedAllFeesAddresses = prevAllFeeAddresses.updated(
+                      event.address,
+                      Set(snapshotFeesInfo.ownerAddress, snapshotFeesInfo.stakingAddress).flatten
+                    )
```

Comment:
I am wondering if we shouldn't concat it with the previous `Set` instead of overriding it in every iteration.

```scala
val updatedAllFeesAddresses = prevAllFeeAddresses.updatedWith(event.address) { existing => 
   val added = Set(snapshotFeesInfo.ownerAddress, snapshotFeesInfo.stakingAddress).flatten
   existing.map(_ ++ added).orElse(added.some)
}
```

Reply:
I mean, we probably should accumulate addresses every iteration. Because previously we were only keeping 2 last addresses